src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractMethodTests.cs

```diff
@@ -2096,5 +2096,570 @@ private static System.Range NewMethod()
+
+class C
+{
+    public string M()
+    {
+        string? a = string.Empty;
+        string? b = string.Empty;
+        string c = {|Rename:NewMethod|}(ref a, ref b);
+        return c;
+    }
+
+    private static string NewMethod(ref string a, ref string b)
+    {
+        var c = a + b;
+        a = string.Empty;
+        c += a;
+        b = ""test"";
+        c = a + b + c;
+        return c;
+    }
+}");
+
+        // There's a case below where flow state correctly asseses that the variable
+        // 'x' is non-null when returned. It's wasn't obvious when writing, but that's 
+        // due to the fact the line above it being executed as 'x.ToString()' would throw
+        // an exception and the return statement would never be hit. The only way the return
+        // statement gets executed is if the `x.ToString()` call succeeds, thus suggesting 
+        // that the value is indeed not null.
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsExtractMethod)]
+        public Task TestFlowNullableReturn_NotNull1()
+            => TestInRegularAndScriptAsync(
+@"#nullable enable
+
+class C
+{
+    public string? M()
+    {
+        [|string? x = null;
+        x.ToString();|]
+
+        return x;
+    }
+}",
+@"#nullable enable
+
+class C
+{
+    public string? M()
+    {
+        string? x = {|Rename:NewMethod|}();
+
+        return x;
+    }
+
+    private static string NewMethod()
+    {
+        string? x = null;
+        x.ToString();
+        return x;
+    }
+}");
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsExtractMethod)]
+        public Task TestFlowNullableReturn_NotNull2()
+            => TestInRegularAndScriptAsync(
+@"#nullable enable
+
+class C
+{
+    public string? M()
+    {
+        [|string? x = null;
+        x?.ToString();
+        x = string.Empty;|]
+
+        return x;
+    }
+}",
+@"#nullable enable
+
+class C
+{
+    public string? M()
+    {
+        string? x = {|Rename:NewMethod|}();
+
+        return x;
+    }
+
+    private static string NewMethod()
+    {
+        string? x = null;
+        x?.ToString();
+        x = string.Empty;
+        return x;
+    }
+}");
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsExtractMethod)]
+        public Task TestFlowNullable_Lambda()
+            => TestInRegularAndScriptAsync(
+@"#nullable enable
+
+using System;
+
+class C
+{
+    public string? M()
+    {
+        [|string? x = null;
+        Action modifyXToNonNull = () =>
+        {
+            x += x;
+        };
+
+        modifyXToNonNull();|]
+
+        return x;
+    }
+}",
+@"#nullable enable
+
+using System;
+
+class C
+{
+    public string? M()
+    {
+        string? x = {|Rename:NewMethod|}();
+
+        return x;
+    }
+
+    private static string? NewMethod()
+    {
+        string? x = null;
+        Action modifyXToNonNull = () =>
+        {
+            x += x;
+        };
+
+        modifyXToNonNull();
+        return x;
+    }
+}");
     }
```

Comment:
Do we have any tests that handle the async case? So if something extracts as `Task<string?>` but we realize we realize all the returns are just "string" we also drop the `?`? Feel free to file a deferred item for this assuming we don't do something bad.

Reply:
#38127 to cover nested nullability not working